external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -2327,36 +2327,123 @@ fn parse_function_decl(
     start_loc: usize,
     modifiers: Modifiers,
     context: &mut Context,
-) -> Result<Function, Box<Diagnostic>> {
+) -> Result<(Function, Option<Box<Diagnostic>>), Box<Diagnostic>> {
     let Modifiers {
         visibility,
         entry,
         native,
         macro_,
     } = modifiers;
 
+    // certain portions of of the partially constructed function must be updated on return
+    fn update_partial(start_loc: usize, context: &mut Context, mut result: Function) -> Function {
+        result.loc = make_loc(
+            context.tokens.file_hash(),
+            start_loc,
+            context.tokens.previous_end_loc(),
+        );
+        result.body = sp(context.tokens.current_token_loc(), FunctionBody_::Native);
+        result
+    }
+
     // "fun" <FunctionDefName>
     consume_token(context.tokens, Tok::Fun)?;
-    let name = FunctionName(parse_identifier(context)?);
-    let type_parameters = parse_optional_type_parameters(context)?;
+    let nid = parse_identifier(context)?;
+    let name = FunctionName(nid);
+
+    let loc = make_loc(
+        context.tokens.file_hash(),
+        start_loc,
+        context.tokens.previous_end_loc(),
+    );
+    // create a largely empty function
+    let mut result = Function {
+        attributes,
+        loc,
+        visibility: visibility.unwrap_or(Visibility::Internal),
+        entry,
+        macro_,
+        signature: FunctionSignature {
+            type_parameters: vec![],
+            parameters: vec![],
+            return_type: sp(name.loc(), Type_::Unit),
+        },
+        name,
+        body: sp(context.tokens.current_token_loc(), FunctionBody_::Native),
+    };
```

Comment:
A rather nit-picky nit. But thoughts on making a local builder here? 

Since there aren't that many fields we would need to set, we could do:
```
let mut type_parameters = None;
let mut parameters = None;
let mut return_type = None;
let mut body = None;
let mk_result = |current_loc| Function {
    loc,
    visibility: visibility.unwrap_or(Visibility::Internal),
    entry,
    macro_,
    signature: FunctionSignature {
        type_parameters: type_parameters.unwrap_or_default(),
        parameters: parameters.unwrap_or_default(),
        return_type: return_type.unwrap_or_else(|| sp(name.loc(), Type_::UnresolvedError)),
    },
    name,
    body: body.unwrap_or_else(|| sp(current_loc, FunctionBody_::UnresolvedError)),
  };

Reply:
It's not _that_ much different, but feels a lot more intentional, which makes me happy :)

Reply:
Done! Though I did turn it to a macro as it looks like Rust does not allow to modify variables after they are captured by the closure.